### PR TITLE
Have roc survival functions return things grouped by `.eval_time`

### PR DESCRIPTION
**Pull request by EmilHvitfeldt**
Still left:

- [x] update `autoplot()` for `roc_curve_survival()`
- [x] update ducumentation to reflect new output format

To close https://github.com/tidymodels/yardstick/issues/392 and close https://github.com/tidymodels/yardstick/issues/391

``` r
library(yardstick)

res <- roc_curve_survival(
  lung_surv,
  truth = surv_obj,
  .pred
)

res
#> # A tibble: 650 × 4
#>    .threshold sensitivity specificity .eval_time
#>         <dbl>       <dbl>       <dbl>      <dbl>
#>  1   -Inf          0            1            100
#>  2      0.615      0            0.995        100
#>  3      0.723      0.0334       0.985        100
#>  4      0.727      0.133        0.985        100
#>  5      0.730      0.167        0.985        100
#>  6      0.732      0.233        0.980        100
#>  7      0.739      0.233        0.969        100
#>  8      0.741      0.267        0.964        100
#>  9      0.746      0.267        0.959        100
#> 10      0.748      0.267        0.954        100
#> # ℹ 640 more rows

roc_auc_survival(
  lung_surv,
  truth = surv_obj,
  .pred
)
#> # A tibble: 5 × 4
#>   .metric          .estimator .eval_time .estimate
#>   <chr>            <chr>           <dbl>     <dbl>
#> 1 roc_auc_survival standard          100     0.659
#> 2 roc_auc_survival standard          200     0.679
#> 3 roc_auc_survival standard          300     0.688
#> 4 roc_auc_survival standard          400     0.648
#> 5 roc_auc_survival standard          500     0.662

library(ggplot2)

res |>
  autoplot()
```

<sup>Created on 2023-04-10 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### CODE REVIEW

File: R/surv-roc_survival_curve.R [hunk 1]
```diff
@@ -140,4 +164,27 @@ roc_curve_survival_impl <- function(truth,
...
+  # create aesthetic
+  roc_aes <- ggplot2::aes(
+    x = 1 - specificity,
+    y = sensitivity,
+    color = .eval_time
+  )
+
+  # build the graph
+  roc_chart <- roc_chart %+%
+    ggplot2::geom_path(mapping = roc_aes) %+%
```
Comment (topepo) [inline R/surv-roc_survival_curve.R:183]: Use `geom_step(direction = "hv")` instead of a line.

File: R/surv-roc_survival_curve.R [hunk 2]
```diff
@@ -140,4 +164,27 @@ roc_curve_survival_impl <- function(truth,
...
+    ggplot2::geom_abline(lty = 3) %+%
+    ggplot2::coord_equal() %+%
+    ggplot2::theme_bw() %+%
+    ggplot2::scale_color_viridis_c()
```
Comment (topepo) [inline R/surv-roc_survival_curve.R:187]: Let the user pick the color palette and theme

File: R/surv-roc_survival_curve.R [hunk 3]
```diff
@@ -7,7 +7,7 @@
 #'
 #' @return
 #' A tibble with class `roc_survival_df`, `grouped_roc_survival_df` having
-#' columns `.threshold`, `recall`, and `precision`.
+#' columns `.threshold`, `recall`, `precision`, and `.eval_time`.
```
Comment (topepo) [inline R/surv-roc_survival_curve.R:10]: I think that this is a copy/paste error (or it is being computed completely wrong)

File: R/surv-roc_survival_curve.R [hunk 4]
```diff
@@ -132,6 +154,8 @@ roc_curve_survival_impl <- function(truth,
   )
   specificity <- cumsum(specificity)
   specificity <- specificity / specificity_denom
+  specificity <- dplyr::if_else(specificity > 1, 1, specificity)
```
Comment (topepo) [inline R/surv-roc_survival_curve.R:157]: Does this happen?

File: tests/testthat/test-surv-roc_survival_curve.R [hunk 5]
```diff
@@ -7,19 +7,32 @@ test_that("roc_curve_survival works", {
...
+    exp_threshold <- dplyr::filter(exp_threshold, .eval_time == eval_time)
+    exp_threshold <- exp_threshold$.pred_survival
+    exp_threshold <- sort(exp_threshold)
+    exp_threshold <- unique(exp_threshold)
+    exp_threshold <- c(0, exp_threshold, 1)
+    expect_identical(
+      result_tmp$.threshold,
+      exp_threshold
+    )
+
```
Comment (topepo) [inline tests/testthat/test-surv-roc_survival_curve.R:29]: I believe, for this method of estimating the ROC curve, that the curve is monotonically increasing over {1 - spec}. Can you test for that?

File: R/surv-roc_survival_curve.R [hunk 6]
```diff
@@ -98,6 +98,26 @@ roc_curve_survival_impl <- function(truth,
   data <- dplyr::tibble(event_time, delta, estimate)
   data <- tidyr::unnest(data, cols = estimate)
 
+  .eval_times <- unique(data$.eval_time)
+
+  out <- list()
```
Comment (topepo) [inline R/surv-roc_survival_curve.R:103]: should you pre-allocate the list? Maybe use `vector(mode = "list", length = length(.eval_times))`

File: R/surv-roc_survival_curve.R (hunk 2 above)
Comment (topepo) [reply R/surv-roc_survival_curve.R:187]: If this code is also in the ROC function, we should get rid of it. I've been frustrated by this in the past.

**Review by topepo — COMMENTED** [review]
Are we using the benchmarks from the other package to test with? Even for early times, this would be a good idea. 

Even writing out the sensitivity and specificity code and testing on a few thresholds/eval times is a good idea.

File: R/surv-roc_survival_curve.R (hunk 4 above)
Comment (EmilHvitfeldt) [reply R/surv-roc_survival_curve.R:157]: the changes in this PR made it so the last element of `specificity` for `eval_time = 200` was `2*10^-12` more than 1...

File: tests/testthat/test-surv-roc_survival_curve.R (hunk 5 above)
Comment (EmilHvitfeldt) [reply tests/testthat/test-surv-roc_survival_curve.R:29]: Correct me if I'm wrong, but isn't this what this is testing: 

https://github.com/tidymodels/yardstick/pull/407/files/7a6e54cac3fb33fecd656dc08ca48d362a49f438#diff-a74339885fdf4dd311c9ca9747ea3bdc09ebb5db88b09039172dba65877af82eR34-R36

File: R/surv-roc_survival_curve.R (hunk 6 above)
Comment (EmilHvitfeldt) [reply R/surv-roc_survival_curve.R:103]: The way we are using it, I don't think it matters too much

``` r
data(ames, package = "modeldata")

f1 <- function(n) {
  out <- list()
  
  for (i in seq_len(n)) {
    out[[i]] <- ames
  }
  out
}

f2 <- function(n) {
  out <- vector(mode = "list", length = length(n))
  
  for (i in seq_len(n)) {
    out[[i]] <- ames
  }
  out
}

bench::mark(
  f1(5),
  f2(5)
)
#> # A tibble: 2 × 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 f1(5)        4.26µs   4.39µs   217895.    7.14MB      0  
#> 2 f2(5)         4.3µs   4.51µs   216270.  160.56KB     21.6

bench::mark(
  f1(500),
  f2(500)
)
#> # A tibble: 2 × 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 f1(500)       442µs    455µs     2177.    80.3KB     8.25
#> 2 f2(500)       442µs    456µs     2186.    80.3KB     6.14
```

<sup>Created on 2023-04-07 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Comment (EmilHvitfeldt) [reply R/surv-roc_survival_curve.R:103]: lists appear to not care about this as much as numerics do

``` r
f1 <- function(n) {
  out <- numeric()
  
  for (i in seq_len(n)) {
    out[[i]] <- i
  }
  out
}

f2 <- function(n) {
  out <- numeric(n)
  
  for (i in seq_len(n)) {
    out[[i]] <- i
  }
  out
}

bench::mark(
  f1(5),
  f2(5)
)
#> # A tibble: 2 × 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 f1(5)         902ns    984ns   912421.    4.18MB       0 
#> 2 f2(5)         533ns    615ns  1516210.   26.53KB     152.

bench::mark(
  f1(500),
  f2(500)
)
#> # A tibble: 2 × 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 f1(500)      50.6µs   57.2µs    16807.   80.28KB    67.1 
#> 2 f2(500)        11µs   11.4µs    85729.    3.95KB     8.57
```

File: R/surv-roc_survival_curve.R (hunk 1 above)
Comment (EmilHvitfeldt) [reply R/surv-roc_survival_curve.R]: done in [`0826829` (#407)](https://github.com/tidymodels/yardstick/pull/407/commits/0826829a016e37a5d0f7c55073b52da45868cc07)

File: R/surv-roc_survival_curve.R (hunk 2 above)
Comment (EmilHvitfeldt) [reply R/surv-roc_survival_curve.R]: done in [`3cc507c` (#407)](https://github.com/tidymodels/yardstick/pull/407/commits/3cc507cf0864667591934b7d5c3c3f37713df9b2)

File: R/surv-roc_survival_curve.R (hunk 3 above)
Comment (EmilHvitfeldt) [reply R/surv-roc_survival_curve.R]: done in [`0153bb4` (#407)](https://github.com/tidymodels/yardstick/pull/407/commits/0153bb437cf3b688eafb7f6636ac4256c781cbf8)

**Comment by topepo** [comment]
One other thing: we might want to make `.eval_time` non-numeric format in `autopilot()`. Maybe use `format(.eval_time)` before plotting?

**Pull request merged**